File: src/oidc/code_request.rs
use super::code_state::get_state;
use crate::auth::password::calculate_password_hash;
use crate::crypto::rand::random_str;
use crate::data::oidc_flow::code_request::CodeRequest;
use crate::data::oidc_flow::code_response::CodeResponse;
use crate::data::oidc_relying_party::RelyingParty;

pub fn code_request(req: CodeRequest) -> Result<CodeResponse, ()> {
    let relying_party = RelyingParty::find(&req.client_id);
    if relying_party.is_none() {
        dbg!("invalid");
        return Err(());
    }
    let relying_party = relying_party.unwrap();

    if req.client_secret.is_none() {
        dbg!("invalid");
        return Err(());
    }

    if calculate_password_hash(&req.client_secret.unwrap(), &req.client_id)
        != relying_party.hashed_client_secret
    {
        dbg!("invalid");
        return Err(());
    }

    let saved_state = get_state(&req.code);
    if saved_state.is_none() {
        dbg!("invalid");
        return Err(());
    }
    let saved_state = saved_state.unwrap();

    if req.client_id != saved_state.client_id {
        dbg!("invalid");
        return Err(());
    }

    if req.redirect_uri != saved_state.redirect_uri {
        dbg!("invalid");
        return Err(());
    }

    if req.grant_type != "authorization_code" {
        dbg!("invalid");
        return Err(());
    }

    Ok(CodeResponse {
        access_token: random_str(16),
        token_type: "bearer".to_string(),
        id_token: saved_state.id_token,
        scope: saved_state.scope,
    })
}

File: internal/ceremony/code_request.go
package ceremony

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	"github.com/comame/accounts.comame.xyz/internal/auth"
	"github.com/comame/accounts.comame.xyz/internal/db"
	"github.com/comame/accounts.comame.xyz/internal/kvs"
	"github.com/comame/accounts.comame.xyz/internal/oidc"
)

func HandleCodeRequest(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("Content-Type", "application/json;charset=UTF-8")
	w.Header().Add("Cache-Control", "no-store")
	w.Header().Add("Pragma", "no-cache")

	if err := r.ParseForm(); err != nil {
		responseError(w, messageBadRequest)
		return
	}
	req := oidc.CodeRequest{
		ClientID:     r.Form.Get("client_id"),
		ClientSecret: r.Form.Get("client_secret"),
		GrantType:    r.Form.Get("grant_type"),
		Code:         r.Form.Get("code"),
		RedirectURI:  r.Form.Get("redirect_uri"),
	}

	// client_secret_basic
	clientID, clientSecret, ok := r.BasicAuth()
	if ok && (req.ClientID != "" || req.ClientSecret != "") {
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, `{ "error": "invalid_request" }`)
		return
	}
	if ok {
		req.ClientID = clientID
		req.ClientSecret = clientSecret
	}

	res, err := handleCodeRequestInternal(req)
	if err != nil {
		fmt.Println(err)
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, `{ "error": "invalid_request" }`)
		return
	}

	j, err := json.Marshal(res)
	if err != nil {
		fmt.Println(err)
		w.WriteHeader(http.StatusBadRequest)
		io.WriteString(w, `{ "error": "invalid_request" }`)
		return
	}

	if res.Error != "" {
		w.WriteHeader(http.StatusBadRequest)
		w.Write(j)
		return
	}

	w.Write(j)
}

func handleCodeRequestInternal(r oidc.CodeRequest) (*oidc.CodeResponse, error) {
	rp, err := db.RelyingParty_select(r.ClientID)
	if err != nil && errors.Is(err, sql.ErrNoRows) {
		return &oidc.CodeResponse{
			Error: "invalid_client",
		}, nil
	}
	if err != nil {
		return nil, err
	}

	if auth.CalculatePasswordHash(r.ClientSecret, r.ClientID) != rp.HashedClientSecret {
		return &oidc.CodeResponse{
			Error: "invalid_client",
		}, nil
	}

	state, err := kvs.CodeState_get(r.Code)
	if err != nil {
		return &oidc.CodeResponse{
			Error: "invalid_grant",
		}, nil
	}

	kvs.CodeState_delete(r.Code)

	if r.ClientID != state.Aud {
		return nil, errors.New("client_id が保存されたものと違う")
	}

	if r.RedirectURI != state.RedirectURI {
		return nil, errors.New("redirect_uri が保存されたものと違う")
	}

	if r.GrantType != "authorization_code" {
		return nil, errors.New("grant_type が authorization_code ではない")
	}

	at, exp, err := createAccessToken(state.Sub, state.Scope)
	if err != nil {
		return nil, err
	}

	return &oidc.CodeResponse{
		AccessToken: at,
		TokenType:   "Bearer",
		ExpiresIn:   exp,
		Scope:       state.Scope,
		IDToken:     state.IDToken,
	}, nil
}
